api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,629 @@
 package aws
 
-import "fmt"
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"sort"
+	"strconv"
+
+	"github.com/zpatrick/go-bytesize"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/autoscaling"
+	"github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/quintilesims/layer0/api/job"
+	"github.com/quintilesims/layer0/api/provider"
+	awsc "github.com/quintilesims/layer0/common/aws"
+	"github.com/quintilesims/layer0/common/config"
+	"github.com/quintilesims/layer0/common/errors"
+	"github.com/quintilesims/layer0/common/models"
+)
 
 type EnvironmentScaler struct {
+	Client              *awsc.Client
+	EnvironmentProvider provider.EnvironmentProvider
+	ServiceProvider     provider.ServiceProvider
+	TaskProvider        provider.TaskProvider
+	JobStore            job.Store
+	Config              config.APIConfig
+	deployCache         map[string][]models.ResourceConsumer
 }
 
-func NewEnvironmentScaler() *EnvironmentScaler {
-	return &EnvironmentScaler{}
+func NewEnvironmentScaler(a *awsc.Client, e provider.EnvironmentProvider, s provider.ServiceProvider, t provider.TaskProvider, j job.Store, c config.APIConfig) *EnvironmentScaler {
+	return &EnvironmentScaler{
+		Client:              a,
+		EnvironmentProvider: e,
+		ServiceProvider:     s,
+		TaskProvider:        t,
+		JobStore:            j,
+		Config:              c,
+	}
 }
 
 func (e *EnvironmentScaler) Scale(environmentID string) error {
-	return fmt.Errorf("EnvironmentScaler not implemented")
+	// GET RESOURCE PROVIDERS
+	clusterName := addLayer0Prefix(e.Config.Instance(), environmentID)
+
+	resourceProviders, err := e.getResourceProviders(clusterName)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] resourceProviders for env '%s': %#v", environmentID, resourceProviders)
+
+	// GET RESOURCE CONSUMERS
+	resourceConsumers, err := e.getResourceConsumers(clusterName)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] resourceConsumers for env '%s': %#v", environmentID, resourceConsumers)
+
+	info, err := e.scale(clusterName, resourceProviders, resourceConsumers)
+	if err != nil {
+		return err
+	}
+	fmt.Println(info)
+	return nil
+}
+
+func (e *EnvironmentScaler) scale(clusterName string, providers []*models.ResourceProvider, consumers []models.ResourceConsumer) (*models.ScalerRunInfo, error) {
+	scaleBeforeRun := len(providers)
+	var errs []error
+
+	// check if we need to scale up
+	for _, consumer := range consumers {
+		hasRoom := false
+
+		// first, sort by memory so we pack tasks by memory as tightly as possible
+		sortProvidersByMemory(providers)
+
+		// next, place any unused providers in the back of the list
+		// that way, we can can delete them if we avoid placing any tasks in them
+		sortProvidersByUsage(providers)
```

Comment:
I may be misunderstanding how the sort package works with this, so please correct me if I'm wrong. If you sort the same object (the length and members are unchanged after the first sort) two consecutive times, mathematically speaking, wouldn't the second sort just simply override what the first sort did?  These functions don't appear to manipulate (add or remove) any items into the slice that I can see, just simply sorting them. Assuming the length and members of `providers` is unchanged after the first sort, I don't think the second sort operation cares what the first sort did.

Reply:
I'm implying here that only `sortProvidersByUsage` is required.

Reply:
@jparsons04 `sortProvidersByMemory` uses an integer comparison, `sortProvidersByUsage` just does a boolean comparison. Doing these sorts allows unused providers to bubble to the end of the list - e.g. it's the last candidate we should try to place a resource in. 

This is important because if we have unused providers at the end of this process, we know we can scale down.

Reply:
Ok, then that makes sense if the sorts are operating on numeric then boolean comparisons.